src/codegen/codegen.cc

```diff
@@ -58,20 +60,122 @@ runtime::Module Build(const Array<LoweredFunc>& funcs,
+    // we will not produce import_tree_.
+    bool has_import_tree = true;
+    if (IsDSOModule(mod_->type_key()) && mod_->imports().empty()) {
+      has_import_tree = false;
+    }
+    uint64_t sz = 0;
+    if (has_import_tree) {
+      // we will append one key for _import_tree
+      // The layout is the same as before: binary_size, key, logic, key, logic...
+      sz = mod_vec_.size() + 1;
+    } else {
+      // Keep the old behaviour
+      sz = mod_->imports().size();
+    }
+    stream->Write(sz);
+
+    for (auto m : mod_vec_) {
+      std::string mod_type_key = m->type_key();
+      if (!IsDSOModule(mod_type_key)) {
+        stream->Write(mod_type_key);
+        m->SaveToBinary(stream);
+      } else {
+        if (has_import_tree) {
+          mod_type_key = "_lib";
+          stream->Write(mod_type_key);
+        }
+      }
+    }
+
+    // Write _import_tree key if we have
+    if (has_import_tree) {
+      std::string import_key = "_import_tree";
+      stream->Write(import_key);
+      stream->Write(import_tree_);
+    }
+  }
+
+ private:
+  void Init() {
+    CreateModuleIndex();
+    CreateImportTree();
+  }
+
+  void CreateModuleIndex() {
+    std::unordered_set<const runtime::ModuleNode*> visited {mod_.operator->()};
+    std::vector<runtime::ModuleNode*> stack {mod_.operator->()};
+    uint64_t module_index = 0;
+
+    while (!stack.empty()) {
+      runtime::ModuleNode* n = stack.back();
+      stack.pop_back();
+      mod2index_[n] = module_index++;
+      mod_vec_.emplace_back(n);
+      for (runtime::Module m : n->imports()) {
+        runtime::ModuleNode* next = m.operator->();
+        if (visited.count(next)) {
+          continue;
+        }
+        visited.insert(next);
+        stack.push_back(next);
+      }
+    }
+    import_tree_.resize(mod_vec_.size());
+  }
+
+  void CreateImportTree() {
+    std::vector<uint64_t> csr_row_ptr {0};
+    std::vector<uint64_t> csr_col_indices;
+    std::vector<uint64_t> csr_values;
+    for (auto m : mod_vec_) {
+      for (size_t i = 0; i < m->imports().size(); i++) {
+        runtime::Module module = m->imports()[i];
+        uint64_t mod_index = mod2index_[module.operator->()];
+        csr_values.push_back(mod_index);
+        csr_col_indices.push_back(i);
+      }
+      csr_row_ptr.push_back(csr_values.size());
+      import_tree_[mod2index_[m]].resize(m->imports().size());
+    }
+
+    for (size_t i = 0; i < mod_vec_.size(); i++) {
+      for (size_t j = csr_row_ptr[i]; j < csr_row_ptr[i + 1]; j++) {
+        import_tree_[i][csr_col_indices[j]] = csr_values[j];
+      }
+    }
+  }
+
+  bool IsDSOModule(const std::string& key) {
+    return key == "llvm" || key == "c";
+  }
+
+  runtime::Module mod_;
+  // construct module to index
+  std::unordered_map<runtime::ModuleNode*, size_t> mod2index_;
+  // index -> module
+  std::vector<runtime::ModuleNode*> mod_vec_;
+  std::vector<std::vector<uint64_t>> import_tree_;
```

Comment:
we can directly store the csr import tree and serialize it, instead of using the import_tree in the vector of vector